File: code/159/solution.go
package main /*{{{*/

import (
	"fmt"
)

const (
	e4  = int64(10000)
	e5  = int64(100000)
	e6  = int64(1000000)
	e7  = int64(10000000)
	e8  = e4 * e4
	e9  = e8 * 10
	mod = int64(1e9 + 7)
)

/*}}}*/
var ans = [e6]int{}

func cal(a int) int {
	ret := (a-1)%9 + 1
	for i := 2; i*i <= a; i++ {
		if a%i == 0 && ans[i]+ans[a/i] > ret {
			ret = ans[i] + ans[a/i]
		}
	}
	ans[a] = ret
	return ret
}
func main() {
	ans := 0
	for i := 2; i < int(e6); i++ {
		ans += cal(i)
	}
	fmt.Println(ans)
}

File: code/159/solution.cpp
#include "../template/bigint.cpp"
#include "../template/china_reminder.cpp"
#include "../template/euler.cpp"
#include "../template/pell.cpp"
#include "../template/prime.cpp"
#include "../template/prime_cnt.cpp"
#include <iostream>/*{{{*/
#include <map>
#include <math.h>
#include <set>
#include <vector>
const long long e2 = 100;
const long long e3 = e2 * 10;
const long long e4 = e3 * 10;
const long long e5 = e4 * 10;
const long long e6 = e4 * e2;
const long long e7 = e6 * 10;
const long long e8 = e7 * 10;
const long long e9 = e8 * 10;
long long _mod     = e9 + 7;
using namespace std; /*}}}*/

int ans[e6];
int cal(int a) {
    int ret = (a - 1) % 9 + 1;
    for (int i = 2; i * i <= a; i++) {
        if (a % i == 0) {
            ret = max(ret, ans[i] + ans[a / i]);
        }
    }
    ans[a] = ret;
    return ret;
}
int main() {
    int ans = 0;
    for (int i = 2; i < e6; i++) {
        ans += cal(i);
    }
    cout << cal(24) << endl;
    cout << ans << endl;
    return 0;
}
